configuration_go/schemas/thanos/cache/cache.go

```diff
@@ -0,0 +1,17 @@
+package cache
+
+type IndexCacheProvider string
+
+// Taken from github.com/thanos-io/pkg/store/cache/factory.go
+
+const (
+	INMEMORY  IndexCacheProvider = "IN-MEMORY"
+	MEMCACHED IndexCacheProvider = "MEMCACHED"
+	REDIS     IndexCacheProvider = "REDIS"
+)
+
+// IndexCacheConfig specifies the index cache config.
+type IndexCacheConfig struct {
+	Type   IndexCacheProvider `yaml:"type"`
+	Config interface{}        `yaml:"config"`
```

Comment:
Is there a way to avoid the empty interface type here and use something else? Maybe generics to state it could be `MemcachedClientConfig | RedisClientConfig | MemoryCacheConfig`?

Reply:
So the best I could come up with is something like that:

```Go
type cacheConfig interface {
	redis.RedisClientConfig | memcached.MemcachedClientConfig | memory.MemoryCacheConfig
	Type() string
}

// IndexCacheConfig specifies the index cache config.
type IndexCacheConfig[T cacheConfig] struct {
	configType string `yaml:"type"`
	config     T      `yaml:"config"`
}

func NewConfig[T cacheConfig](c T) *IndexCacheConfig[T] {
	return &IndexCacheConfig[T]{
		configType: c.Type(),
		config:     c,
	}
}
```

Is that what you are looking for? If so I can change with that.